spec/s3_spec.rb

```diff
@@ -44,4 +44,91 @@
       it { expect{S3.ensure_bucket('bucket')}.to raise_error Aws::S3::Errors::Forbidden}
     end
   end
+
+  describe '#upload' do
+    it 'throws an error without a filename' do
+      expect {
+        S3.upload('bucket')
+      }.to raise_error "Must provide filename to S3.upload()"
```

Comment:
If a filename is always required, does it make sense to make it an argument instead of having it part of opts?